### PR TITLE
Outline use of type_annotation_map to fix mypy issues

### PR DESCRIPTION
**Pull request by jmchilton**
Builds on #17897 

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/model/__init__.py
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: My strong suggestion is to not include `Optional` here. In the model definition, `Optional` will be used to infer nullability of the database field:
`foo: Mapped[mytype]` and `foo: Mapped[Optional[mytype]]` will generate different database schemas: the former will define foo as `NOT NULL`. Furthermore, this can be overwritten with the `nullable` attribute in `mapped_column`: `foo: Mapped[mytype] = mapped_column(nullable=True)` will generate a definition for a nullable database field (while allowing `None` in the Python app). And vise-versa.
So, seeing or not seeing `Optional` in the model definition is really important.

File: lib/galaxy/model/__init__.py
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]:
```suggestion
    transform: Mapped[Optional[TRANSFORM_ACTIONS]] = mapped_column(MutableJSONType)
```
Otherwise this field will be defined as `NOT NULL` in the database (when loaded from the model definition)

File: lib/galaxy/model/__init__.py
Comment (jmchilton) [reply lib/galaxy/model/__init__.py]: I think it respects the runtime type description and isn't like looking for Optional using raw Python source code reading or something naive like that. 

``create_test.py`` -->

```python
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.schema import CreateTable
from typing import Optional

OptionalFoo = Optional[str]
RequiredFoo = str


class Base(DeclarativeBase):
    pass


class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    optional_foo: Mapped[OptionalFoo]
    required_foo: Mapped[RequiredFoo]


print(CreateTable(SomeClass.__table__))
```


```
> python create_test.py 

CREATE TABLE some_table (
	id INTEGER NOT NULL, 
	name VARCHAR NOT NULL, 
	optional_foo VARCHAR, 
	required_foo VARCHAR NOT NULL, 
	PRIMARY KEY (id)
)
```

Comment (jdavcs) [reply lib/galaxy/model/__init__.py]: Oh, absolutely. I didn't mean it wouldn't work - I was only referring to keeping the definition explicit and straightforward. Given our 10K+ lines model definition module, the line defining our custom `OptionalFoo` type can be thousands of lines removed from where it's used. The use of "optional" as a prefix for such a type name would be our own naming convention that we are bound to eventually disregard. I think, having Python's `typing.Optional` in the model's definition' makes it more explicit. Given that this one term determines what goes into the database schema, I think it would be better to keep it where the field is defined - i.e., make it less error prone and make the model easier to understand.

Comment (jmchilton) [reply lib/galaxy/model/__init__.py]: I don't agree really but I've made the change.

Comment (jdavcs) [reply lib/galaxy/model/__init__.py]: Thank you, I do appreciate it! We can revisit this after we have all the type/model improvements in place.

File: lib/galaxy/model/__init__.py
Comment (jdavcs) [inline lib/galaxy/model/__init__.py:221]:
```suggestion
        STR_TO_STR_DICT: JSONType,
        TRANSFORM_ACTIONS: MutableJSONType,
```
We use `Optional` in the model definition

**Comment by jdavcs** [comment]
Since we've decided to keep `HasTable`, the second commit (82fd47bc3e4ece4956e157f4784b378b41fef423) needs to be dropped. I can do that and push to your branch and merge if that's OK.

Also, to make `DeclarativeBase` work, we need this commit: dcf2d843049adedf0c7ba56578aad332d3557b95, but that is part of #17922, which I'll merge today.

**Review by jdavcs — APPROVED** [review]
Thanks @jmchilton!

(all: please don't merge before 82fd47bc3e4ece4956e157f4784b378b41fef423 commit is dropped)

File: lib/galaxy/model/__init__.py
Comment (jmchilton) [reply lib/galaxy/model/__init__.py:221]: Well my thinking was the thing in map needed to match the column definition completely. If it doesn't have to match the column definition exactly... I think we don't actually need this map right? I am thinking we just drop this if it isn't actually needed. We would have test failures if this was needed and wrong.

Comment (jdavcs) [reply lib/galaxy/model/__init__.py:221]: @jmchilton it works as intended; I've reread the docs and tested it to make sure I'm not misreading stuff. Here's how I read it:
```py
STR_TO_STR_DICT = Dict[str, str]  # we define a type alias
...
STR_TO_STR_DICT: JSONType  
# in the statement above (type annotation map), we say that the Python type STR_TO_STR_DICT 
# should be mapped to SQLAlchemy's JSONType, which is a SQLAlchemy type that knows how to 
# generate its DDL. 
# If we substituted JSONType for Integer, the Python code would still expect Dict[str, str],
# but in the db that field would be defined as an integer type.
...
object_store_id_overrides: Mapped[Optional[STR_TO_STR_DICT]]  # that's all we need here!
# the field will be typed as Optional[Dict[str, str]], represented as bytea in the database. 
```
The `Optional` type specifier works fine: here's an example that causes a mypy error; removing `Optional` fixes it - as expected:
```py
def test(job: Job):
    job.object_store_id_overrides['key']
test(Job())

# lib/galaxy/model/__init__.py:11430: error: Value of type "Optional[dict[str, str]]" is not indexable  [index]
```

Does this make sense?

(There's a conflict to resolve - I'm happy to take care of this.)

File: lib/galaxy/model/__init__.py
Comment (jdavcs) [inline lib/galaxy/model/__init__.py:1409]:
```suggestion
    object_store_id_overrides: Mapped[Optional[STR_TO_STR_DICT]]
```
I think we don't need this: it's derived from the `type_annotation_map`.
Same for `transform` below.

File: lib/galaxy/model/__init__.py
Comment (jmchilton) [reply lib/galaxy/model/__init__.py:221]: Maybe this makes sense - I think given that we don't have a typed way to distinguish between what should be mutable or not currently we should drop the dictionary and keep being explicit but I will defer to you until I understand it better.

Update: I mean like we probably have some string to string dictionaries that are mutable and some that are not right? And regardless - we might want to add some or either variety even if we only have one type currently?

Comment (jdavcs) [reply lib/galaxy/model/__init__.py:221]: Yes, i think so (re. your update). Also, that type annotation map may be useful for other type mappings (SA's docs have some examples [here](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#step-five-make-use-of-pep-593-annotated-to-package-common-directives-into-types)), even if the only benefit in those cases would be more succinct code.

**Pull request merged**
